<a href="https://colab.research.google.com/github/sourcecode369/TensorFlow-2.0/blob/master/tensorflow_2.0_docs/TensorFlow%20Core/Tutorials/Text/Simple_RNN_Test_(Return_State_vs_Return_Sequences).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
from __future__ import print_function, division
from builtins import range, input

from keras.models import Model
from keras.layers import Input, LSTM, GRU, RNN
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

T = 8
D = 2
M = 3

In [0]:
X = np.random.randn(1,T, D)
def lstm1():
  input_ = Input(shape=(T, D))
  rnn = LSTM(M, return_state=True)
  x = rnn(input_)
  model = Model(inputs=input_, outputs=x)
  o, h, c = model.predict(X)
  print("o:",o)
  print("h:",h)
  print("c:",c)

def lstm2():
  input_ = Input(shape=(T, D))
  rnn = LSTM(M, return_state=True,return_sequences=True)
  x = rnn(input_)

  model = Model(inputs=input_, outputs=x)
  o, h, c = model.predict(X)
  print("o:",o)
  print("h:",h)
  print("c:",c)

In [0]:
def gru1():
  input_ = Input(shape=(T,D))
  rnn = GRU(M, return_state=True)
  x = rnn(input_)
  model = Model(inputs=input_, outputs=x)
  o, h = model.predict(X)
  print("o:",o)
  print("h:",h)

def gru2():
  input_ = Input(shape=(T,D))
  rnn = GRU(M, return_state=True, return_sequences=True)
  x = rnn(input_)

  model = Model(inputs=input_, outputs=x)
  o, h = model.predict(X)
  print("o:",o)
  print("h:",h)

In [12]:
print("lstm1:")
lstm1()

print("lstm2:")
lstm2()

print("gru1:")
gru1()

print("gru2:")
gru2()

lstm1:
o: [[0.04405436 0.19382982 0.09365804]]
h: [[0.04405436 0.19382982 0.09365804]]
c: [[0.07515235 0.74266243 0.2020666 ]]
lstm2:
o: [[[-0.13985486  0.10550682 -0.09445278]
  [-0.10432478 -0.27001908 -0.00554497]
  [-0.08823357 -0.29882026  0.02738047]
  [-0.00888742 -0.27147183  0.08124414]
  [-0.0464495  -0.15506767  0.01873142]
  [ 0.0823176  -0.07812153  0.0788352 ]
  [ 0.17786328  0.02429481  0.08474788]
  [ 0.19463679 -0.14741603  0.18657811]]]
h: [[ 0.19463679 -0.14741603  0.18657811]]
c: [[ 0.45711023 -0.33413443  0.36481676]]
gru1:
o: [[-0.04083507 -0.62333137 -0.669806  ]]
h: [[-0.04083507 -0.62333137 -0.669806  ]]
gru2:
o: [[[-0.00711141 -0.32353368 -0.2206695 ]
  [-0.21819714 -0.14739999 -0.46059692]
  [-0.22654344 -0.08817856 -0.48051834]
  [-0.18662423  0.01923173 -0.15814897]
  [-0.10439495 -0.14305443 -0.19330898]
  [-0.02846893  0.07500637  0.4005316 ]
  [ 0.11978745  0.22520865  0.6934356 ]
  [ 0.15229619  0.3807733   0.37033397]]]
h: [[0.15229619 0.3807733  0.370

## LSTM (Long Short Term Memory)

#### 1. Return State

For the first LSTM i.e. LSTM 1 the output is the same as the hidden state

    lstm1:
    o: [[0.04405436 0.19382982 0.09365804]]
    h: [[0.04405436 0.19382982 0.09365804]]
    c: [[0.07515235 0.74266243 0.2020666 ]]

#### 2. Return Sequences

For the second LSTM i.e. LSTM 2 the output is longer i.e. 8*3 since the length of the sequence is 8. However the last values of the output state is the same as hidden state.

    lstm2:
    o: [[[-0.13985486  0.10550682 -0.09445278]
      [-0.10432478 -0.27001908 -0.00554497]
      [-0.08823357 -0.29882026  0.02738047]
      [-0.00888742 -0.27147183  0.08124414]
      [-0.0464495  -0.15506767  0.01873142]
      [ 0.0823176  -0.07812153  0.0788352 ]
      [ 0.17786328  0.02429481  0.08474788]
      [ 0.19463679 -0.14741603  0.18657811]]]
    h: [[ 0.19463679 -0.14741603  0.18657811]]
    c: [[ 0.45711023 -0.33413443  0.36481676]]

Thus we can conclude that H and C are the same. In other words, H and C are the hidden state and the cell state of the final time step of the input in an LSTM.




## GRU (Gated Recurrent Unit)

Similarly goes for the GRU. Only there is no cell state gate in an GRU.|